twoliter/src/lock.rs

```diff
@@ -139,6 +195,121 @@ impl Lock {
+            .context("failed to create external-kits directory")?;
+        for image in self.kit.iter() {
+            self.extract_kit(&target_dir, image, arch).await?;
+        }
+        let mut kit_list = Vec::new();
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut kit_list, CanonicalFormatter::new());
+        self.external_kit_metadata()
+            .serialize(&mut ser)
+            .context("failed to serialize external kit metadata")?;
+        write(
+            target_dir.join("external-kit-metadata.json"),
+            kit_list.as_slice(),
+        )
+        .await
+        .context("failed to write external kit metadata")?;
+
+        Ok(())
+    }
+
+    async fn get_manifest(&self, image: &LockedImage, docker_arch: &str) -> Result<ManifestView> {
+        let manifest_bytes = docker!(
+            ["manifest", "inspect", image.source.as_str()],
+            format!("failed to find a kit {}", image.to_string())
+        );
+        let manifest_list: ManifestListView = serde_json::from_slice(manifest_bytes.as_slice())
+            .context("failed to deserialize manifest list")?;
+        manifest_list
+            .manifests
+            .iter()
+            .find(|x| x.platform.as_ref().unwrap().architecture == docker_arch)
+            .cloned()
+            .context(format!(
+                "could not find kit image for architecture '{}' at {}",
+                docker_arch, image.source
+            ))
+    }
+
+    async fn get_index<P>(&self, path: P, digest_uri: &str) -> Result<IndexView>
+    where
+        P: AsRef<Path>,
+    {
+        let oci_archive = docker!(
+            ["save", digest_uri],
+            format!("failed to fetch kit and save to disk from {}", digest_uri)
+        );
+        let mut oci_archive = Archive::new(oci_archive.as_slice());
+        // Unpack the oci image to tempdirectory
+        oci_archive
+            .unpack(path.as_ref())
+            .context("failed to unpack oci image")?;
+        // Now read in the index
+        let index_bytes = read(path.as_ref().join("index.json")).await?;
+        serde_json::from_slice(index_bytes.as_slice())
+            .context("failed to deserialize oci image index")
+    }
+
+    async fn extract_kit<P>(&self, path: P, image: &LockedImage, arch: &str) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let tmp_dir =
+            tempdir().context("failed to create a temporary directory for kit extraction")?;
+        let vendor = image.vendor.clone();
+        let name = image.name.clone();
+        let docker_arch = match arch {
+            "x86_64" => Ok("amd64"),
+            "aarch64" => Ok("arm64"),
+            _ => Err(anyhow!("unsupported architecture '{}'", arch)),
+        }?;
+        let target_path = path.as_ref().join(format!("{vendor}/{name}/{arch}"));
+        create_dir_all(&target_path).await?;
+
+        // First we want to get the correct digest for our architecture
```

Comment:
It would be good to add more comments for this algorithm. It's not really clear where "getting the correct digest" ends, it seems like we're doing other stuff with side effects along the way toward `let digest = ...` (assuming that's where it happens).